src/v2/guide/components.md

```diff
@@ -1,1481 +1,595 @@
+  <button-counter></button-counter>
 </div>
 <script>
-Vue.component('my-component', {
-  template: '<div>A custom component!</div>'
-})
-new Vue({ el: '#example' })
+new Vue({ el: '#components-demo2' })
 </script>
 {% endraw %}
 
-### Local Registration
+Notice that when clicking on the buttons, each one maintains its own, separate `count`. That's because each time you use a component, a new **instance** of it is created.
 
-You don't have to register every component globally. You can make a component available only in the scope of another instance/component by registering it with the `components` instance option:
+### `data` Must Be a Function
 
-``` js
-var Child = {
-  template: '<div>A custom component!</div>'
+When we defined the `<button-counter>` component, you may have noticed that `data` wasn't directly provided an object, like this:
+
+```js
+data: {
+  count: 0
 }
+```
 
-new Vue({
-  // ...
-  components: {
-    // <my-component> will only be available in parent's template
-    'my-component': Child
+Instead, **a component's `data` option must be a function**, so that each instance can maintain an independent copy of the returned data object:
+
+```js
+data: function () {
+  return {
+    count: 0
   }
-})
+}
 ```
 
-The same encapsulation applies for other registerable Vue features, such as directives.
+If Vue didn't have this rule, clicking on one button would affect the data of _all other instances_, like below:
 
-### DOM Template Parsing Caveats
+{% raw %}
+<div id="components-demo3" class="demo">
+  <button-counter2></button-counter2>
+  <button-counter2></button-counter2>
+  <button-counter2></button-counter2>
+</div>
+<script>
+var buttonCounter2Data = {
+  count: 0
+}
+Vue.component('button-counter2', {
+  data: function () {
+    return buttonCounter2Data
+  },
+  template: '<button v-on:click="count++">You clicked me {{ count }} times.</button>'
+})
+new Vue({ el: '#components-demo3' })
+</script>
+{% endraw %}
 
-When using the DOM as your template (e.g. using the `el` option to mount an element with existing content), you will be subject to some restrictions that are inherent to how HTML works, because Vue can only retrieve the template content **after** the browser has parsed and normalized it. Most notably, some elements such as `<ul>`, `<ol>`, `<table>` and `<select>` have restrictions on what elements can appear inside them, and some elements such as `<option>` can only appear inside certain other elements.
+## Organizing Components
 
-This will lead to issues when using custom components with elements that have such restrictions, for example:
+It's common for an app to be organized into a tree of nested components:
 
-``` html
-<table>
-  <my-row>...</my-row>
-</table>
-```
+![Component Tree](/images/components.png)
 
-The custom component `<my-row>` will be hoisted out as invalid content, thus causing errors in the eventual rendered output. A workaround is to use the `is` special attribute:
```

Comment:
This, with the lines above it up to line 98, is what I was referring to- I'm adding this comment in here for quick reference as a talking point :)